go-controller/pkg/ovn/gress_policy.go

```diff
@@ -0,0 +1,193 @@
+package ovn
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/config"
+	v1 "k8s.io/api/core/v1"
+	knet "k8s.io/api/networking/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+type gressPolicy struct {
+	namespace  string
+	name       string
+	policyType knet.PolicyType
+	idx        int
+
+	// peerAddressSet points to the addressSet that holds all peer pod
+	// IP addresess.
+	peerAddressSet *addressSet
+
+	// nsAddressSets holds the names of all namespace address sets
+	nsAddressSets sets.String
+
+	// sortedPeerAddressSets has the sorted peerAddressSets
+	sortedPeerAddressSets []string
+
+	// portPolicies represents all the ports to which traffic is allowed for
+	// the rule in question.
+	portPolicies []*portPolicy
+
+	// ipBlockCidr represents the CIDR from which traffic is allowed
+	// except the IP block in the except, which should be dropped.
+	ipBlockCidr   []string
+	ipBlockExcept []string
+}
+
+type portPolicy struct {
+	protocol string
+	port     int32
+}
+
+func (pp *portPolicy) getL4Match() (string, error) {
+	if pp.protocol == TCP {
+		return fmt.Sprintf("tcp && tcp.dst==%d", pp.port), nil
+	} else if pp.protocol == UDP {
+		return fmt.Sprintf("udp && udp.dst==%d", pp.port), nil
+	} else if pp.protocol == SCTP {
+		return fmt.Sprintf("sctp && sctp.dst==%d", pp.port), nil
+	}
+	return "", fmt.Errorf("unknown port protocol %v", pp.protocol)
+}
+
+func newGressPolicy(policyType knet.PolicyType, idx int, namespace, name string) *gressPolicy {
+	return &gressPolicy{
+		namespace:             namespace,
+		name:                  name,
+		policyType:            policyType,
+		idx:                   idx,
+		nsAddressSets:         sets.String{},
+		sortedPeerAddressSets: make([]string, 0),
+		portPolicies:          make([]*portPolicy, 0),
+		ipBlockCidr:           make([]string, 0),
+		ipBlockExcept:         make([]string, 0),
+	}
+}
+
+func (gp *gressPolicy) ensurePeerAddressSet() error {
+	if gp.peerAddressSet != nil {
+		return nil
+	}
+
+	direction := strings.ToLower(string(gp.policyType))
+	asName := fmt.Sprintf("%s.%s.%s.%d", gp.namespace, gp.name, direction, gp.idx)
+	as, err := NewAddressSet(asName, nil)
+	if err != nil {
+		return err
+	}
+
+	gp.peerAddressSet = as
+	gp.sortedPeerAddressSets = append(gp.sortedPeerAddressSets, as.hashName)
+	sort.Strings(gp.sortedPeerAddressSets)
+	return nil
+}
+
+func (gp *gressPolicy) addPortPolicy(portJSON *knet.NetworkPolicyPort) {
+	protocol := v1.ProtocolTCP
+	if portJSON.Protocol != nil {
+		protocol = *portJSON.Protocol
+	}
+	gp.portPolicies = append(gp.portPolicies, &portPolicy{
+		protocol: string(protocol),
+		port:     portJSON.Port.IntVal,
+	})
+}
+
+func (gp *gressPolicy) addIPBlock(ipblockJSON *knet.IPBlock) {
+	gp.ipBlockCidr = append(gp.ipBlockCidr, ipblockJSON.CIDR)
+	gp.ipBlockExcept = append(gp.ipBlockExcept, ipblockJSON.Except...)
+}
+
+func ipMatch() string {
```

Comment:
I thought @danwinship got rid of this method of detecting ipv6.

Reply:
The NetworkPolicy code isn't dual-stack aware yet. (#1270 was a start, but the approach there didn't pan out)

Reply:
I didn't mean dual stack aware. I remember you made a patch to stop checking for ipv6 using the global config flag. I cannot remember which commit it was, but now I see we use this everywhere:
utilnet "k8s.io/utils/net"
utilnet.IsIPv6String(sourceIP)

Perhaps it would be best to be consistent and use this too on the first address in the set to detect ipv6 or ipv4? Added bonus we don't have to change this code again in the future to support dual stack.

Reply:
We use utilnet in cases where the code is operating on a specific IP address, but in at least some of the cases that use `ipMatch`, the caller is working with an address set, and at the point where we have to decide whether to use `ip.src` or `ipv6.src`, we don't have access to the *contents* of the address set, only its name.

The use of `config.IPv6Mode` here will go away when we make the NetworkPolicy code maintain separate IPv4 and IPv6 address sets for each gressPolicy that needs them.